openprescribing/frontend/tests/price_per_unit/test_savings.py

```diff
@@ -0,0 +1,320 @@
+from collections import defaultdict
+import json
+import warnings
+
+import numpy
+
+from django.core.cache import CacheKeyWarning
+from django.test import TestCase, override_settings
+
+from matrixstore.tests.data_factory import DataFactory
+from matrixstore.tests.matrixstore_factory import (
+    patch_global_matrixstore,
+    matrixstore_from_data_factory,
+)
+from frontend.models import Practice, PCT, Presentation
+from frontend.price_per_unit.substitution_sets import get_substitution_sets
+from frontend.price_per_unit.savings import (
+    CONFIG_MIN_SAVINGS_FOR_ORG_TYPE,
+    CONFIG_TARGET_CENTILE,
+    get_total_savings_for_org,
+)
+
+
+# The dummy cache backend we use in testing warns that our binary cache keys
+# won't be compatible with memcached, but we really don't care
+warnings.simplefilter("ignore", CacheKeyWarning)
+DUMMY_CACHE_SETTING = {
+    "default": {"BACKEND": "django.core.cache.backends.dummy.DummyCache"}
+}
+
+
+@override_settings(CACHES=DUMMY_CACHE_SETTING)
+class PricePerUnitSavingsTest(TestCase):
+    maxDiff = None
+
+    @classmethod
+    def setUpTestData(cls):
+        # Create some sets of substitutable presentations, which each consist
+        # of a generic plus some branded equivalents
+        substitution_sets = []
+        for i in range(4):
+            generic_code = invent_generic_bnf_code(i)
+            brands = invent_brands_from_generic_bnf_code(generic_code)
+            substitution_sets.append([generic_code] + brands)
+
+        # Create some practices and some prescribing for these presentations
+        factory = DataFactory()
+        factory.create_months("2020-01-01", 2)
+        factory.create_practices(10)
+        for bnf_codes in substitution_sets:
+            for bnf_code in bnf_codes:
+                factory.create_presentation(bnf_code=bnf_code)
+        factory.create_prescribing(
+            factory.presentations, factory.practices, factory.months
+        )
+
+        # Copy these objects to the database
+        ccg = PCT.objects.create(name="CCG1", code="ABC", org_type="CCG")
+        for practice in factory.practices:
+            Practice.objects.create(
+                name=practice["name"], code=practice["code"], setting=4, ccg=ccg
+            )
+        for presentation in factory.presentations:
+            Presentation.objects.create(
+                bnf_code=presentation["bnf_code"], name=presentation["name"]
+            )
+
+        cls.substitution_sets = substitution_sets
+        cls.factory = factory
+        cls._remove_patch = patch_global_matrixstore(
+            matrixstore_from_data_factory(factory)
+        )
+        # Clear the cache on this memoized function
+        get_substitution_sets.cache_clear()
```

Comment:
Why is the necessary with a DummyCache backend?

Reply:
This is actually a good illustration of what's confusing about our current caching set up. We've got two sorts of cached values: 1. Things which are held in memory for just the lifetime of the process 2. Things which are written to an external cache backend (previously Redis, soon to be DiskCache).

Caching of type 1 uses `functools.lru_cache` and caching of type 2 uses `django.core.cache`. I was thinking it might be good to have a single unified wrapper which could do either, depending on the arguments used. And then a single place to disable caching for tests. But I haven't quite figured out how that would all work yet.